tests/typecheck/test_request.yml

```diff
@@ -69,3 +69,41 @@
         reveal_type(request.user) # N: Revealed type is "django.contrib.auth.models.User"
     custom_settings: |
         INSTALLED_APPS = ('django.contrib.contenttypes', 'django.contrib.auth')
+
+-   case: request_get_post
+    main: |
+        from django.http.request import HttpRequest
+
+        request = HttpRequest()
+        reveal_type(request)  # N: Revealed type is "django.http.request._MutableHttpRequest"
+        reveal_type(request.GET)  # N: Revealed type is "django.http.request.QueryDict"
+        request.GET['foo'] = 'bar'
+
+        def mk_request() -> HttpRequest:
+            return HttpRequest()
+
+        req = mk_request()
+        reveal_type(req)  # N: Revealed type is "django.http.request.HttpRequest"
+        reveal_type(req.GET)  # N: Revealed type is "django.http.request._ImmutableQueryDict"
+        req.GET.setdefault('foo', 'bar')  # E: This QueryDict is immutable.
+        x = 1  # E: Statement is unreachable
+
+# Will work after merging https://github.com/python/mypy/pull/12572
+-   case: request_get_post_unreachable
+    main: |
+        from django.http.request import HttpRequest
+
+        request = HttpRequest()
+        reveal_type(request)  # N: Revealed type is "django.http.request._MutableHttpRequest"
+        reveal_type(request.GET)  # N: Revealed type is "django.http.request.QueryDict"
+        request.GET['foo'] = 'bar'
+
+        def mk_request() -> HttpRequest:
+            return HttpRequest()
```

Comment:
Sorry, I fail to understand how this works.

None of these calls use `.copy` or `mutable=True`. Why is there any difference?

Reply:
My bad, probably I haven't explained that good enough.

We assume that anything annotated as `HttpRequest` is referring to `{W,A}SGIRequest` and thus has immutable `GET` and `POST`. However, this assumption doesn't hold for `HttpRequest` itself that has mutable dicts. So when we instantiate `HttpRequest` directly, overridden `__new__` returns fake `_MutableHttpRequest` class, so the first part of case is actually operating on it (see `reveal_type` there). The second part works with `HttpRequest` and assumes it's immutable.

Reply:
Do you think that having `_MutableHttpRequest` is worth it? It complicates our code quite a bit.
In my experience, we almost never do such things.

Reply:
@sobolevn For background, the reason of this complexity is not distinguishing between mutable/immutable `QueryDict`, but about what types `QueryDict` methods can return.

1. As streliakov points out in [this comment](https://github.com/typeddjango/django-stubs/pull/899#issuecomment-1086718345), the `dict()` and `__getitem__()` methods of `QueryDict` and `MultiValueDict` may return `str` as well as empty `list` (but never a list with length>0).

   IMO this is crazy behavior on Django's part -- it should throw KeyError -- but nothing we can do about that. This isn't mentioned in official Django documentation, but is mentioned in `MultiValueDict` docstring: https://github.com/django/django/blob/main/django/utils/datastructures.py#L78-L82
4. However, the vast majority of code that uses `QueryDict` is within the request lifecycle of Django. In these cases, the empty lists cannot appear. So `dict()` and `__getitem__()` can safely be assumed to only have `str` values.
5. Large amounts of code rely on the # 2 assumption, even example code from Django's official tutorial ([explained here](https://github.com/typeddjango/django-stubs/pull/909#discussion_r843949589)).
6. The `QueryDict` attributes of Django request are created with `mutable=False`

So there are three options:
* We could have one `QueryDict` class and these methods so they always return `Union[str, List]`. This is the situation after #909. But this causes false positive errors in a significant amount of idiomatic Django code, since most code assumes list return can not appear in `request.GET["key"]`
* Do the opposite, have one `QueryDict` class, and hint these methods as returning `str`. This is **not strictly correct** typing, although the list return is a rare fringe case.
* A mix of the above, which is what this MR does. In Django request lifecycle, use the magic `_ImmutableQueryDict` type, which only holds `str` values. In other cases, assume mutable `QueryDict` may also return empty list for values.

I would prefer the second or third option out of these three.